File: game/systems/remote.js
import * as THREE from 'three';
import { scene } from '../world/scene.js';
import { MP, on, sendGame, netRegistry } from '../net/net.js';
import { entities, makeBar, makeMech, makeTurretEntity, registerEntity, blueBase, redBase, BLUE, RED } from '../entities/entities.js';
import { player } from '../entities/player.js';
import { spawnProjectile, killEntity, damageEntity } from '../entities/projectiles.js';
import { spawnPointFor, teamIndexOf, animateWalk } from '../core/helpers.js';
import { onFlagMsg } from './ctf.js';
import { groundHeightAt } from '../world/world.js';
import { game } from '../core/state.js';
import { showMessage } from '../ui/hud.js';
import { endGame } from '../core/flow.js';

/* ============================================================
   Multiplayer match sync — up to 5 players per team

   Ownership model: each client simulates only what it owns — its
   player mech, the turrets it built, its projectiles. Everyone
   else's entities exist locally as replicas driven by network
   events (the server stamps each relayed event with the sender's
   playerId, so `from` is trustworthy):

     s        15 Hz state (position/yaw/velocity/hp + turret yaws)
     shot     a projectile was fired → spawn a cosmetic copy
     hit      my projectile hit an entity YOU own → you apply it
     hp       authoritative hp echo after a hit was applied
     bhit     damage to a base — bases are shared (unowned), so the
              shooter applies it and everyone else mirrors it
     build    a turret was built
     die      an entity its owner simulates died → mirror it
     respawn  a player redeployed
     fgrab/fdrop/fret/fcap
              capture the flag: what the sender's mech did with a
              flag. Flags are shared and unowned like the bases, so
              only the client simulating that mech reports it and
              everyone else mirrors it (systems/ctf.js)

   Damage to owned entities is shooter-reported but owner-applied,
   so hp has exactly one authority. Base hp converges because every
   client applies every bhit exactly once.
============================================================ */

/* playerId -> {id, name, team, idx, ent, st, connected} */
export const peers = new Map();

if (MP.active) initMatch();

/* floating pilot name so team fights stay readable */
function makeNameTag(text, team) {
  const cv = document.createElement('canvas');
  cv.width = 256; cv.height = 40;
  const ctx = cv.getContext('2d');
  ctx.font = '700 26px system-ui, sans-serif';
  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';
  ctx.shadowColor = 'rgba(0,0,0,0.9)';
  ctx.shadowBlur = 5;
  ctx.fillStyle = team === 'red' ? '#ffb3a6' : '#a9c9ff';
  ctx.fillText(text, 128, 21);
  const tex = new THREE.CanvasTexture(cv);
  const sp = new THREE.Sprite(new THREE.SpriteMaterial({ map: tex, depthTest: false, transparent: true, toneMapped: false }));
  sp.scale.set(9, 9 * 40 / 256, 1);
  sp.renderOrder = 10;
  return sp;
}

function makePeer(p) {
  const idx = teamIndexOf(p.id, p.team, MP.roster);
  const sp = spawnPointFor(p.team, idx);
  const y = groundHeightAt(sp.pos.x, sp.pos.z);
  const model = makeMech(p.team === 'red' ? RED : BLUE);
  model.group.position.set(sp.pos.x, y, sp.pos.z);
  const yaw = Math.atan2(sp.face.x - sp.pos.x, sp.face.z - sp.pos.z);
  const ent = registerEntity({
    kind: 'mech', team: p.team, group: model.group, model, remote: true, owner: p.id,
    netId: `player:${p.id}`,
    hp: 300, maxHp: 300, alive: true,
    hitRadius: 2.4, hitHeight: 7, bar: makeBar(5), barHeight: 8.2,
    yaw, walkPhase: 0, stride: 0, strideF: 1, strideL: 0, y, vy: 0, velX: 0, velZ: 0,
    anchorX: sp.pos.x, anchorZ: sp.pos.z,
  });
  const tag = makeNameTag(p.name, p.team);
  tag.position.y = 9.6;
  ent.group.add(tag);
  peers.set(p.id, {
    id: p.id, name: p.name, team: p.team, idx, ent, connected: true,
    st: { x: sp.pos.x, z: sp.pos.z, y, yaw, vx: 0, vz: 0, moving: false, age: 0 },
  });
}

function initMatch() {
  for (const p of MP.roster) if (p.id !== MP.playerId) makePeer(p);

  on('game', onGameMsg);
  on('peerLeft', (m) => onPeerLeft(m.id, m.name));
  on('peerJoined', (m) => {
    const p = MP.roster.find((r) => r.id === m.id);
    const peer = peers.get(m.id);
    if (!p || (peer && peer.connected)) return;
    makePeer(p); // fresh replica — the rejoining client restarted from scratch
    if (game.state === 'playing') showMessage(`${p.name} RECONNECTED`, '#8ab4ff');
  });
  on('close', () => {
    if (game.state === 'playing') endGame(false, 'CONNECTION TO SERVER LOST');
  });
}

function onPeerLeft(id, name) {
  const peer = peers.get(id);
  if (!peer || !peer.connected) return;
  peer.connected = false;
  // despawn everything the leaver owned — nobody is left to simulate it,
  // and hits on it would never be applied
  for (const e of [...entities]) {
    if (e.owner === id) {
      e.alive = false;
      scene.remove(e.group);
      const i = entities.indexOf(e);
      if (i >= 0) entities.splice(i, 1);
    }
  }
  if (game.state !== 'playing') return;
  showMessage(`${name || peer.name} DISCONNECTED`, peer.team === MP.myTeam ? '#ff8a7a' : '#8ab4ff');
  if (peer.team === MP.enemyTeam
    && ![...peers.values()].some((o) => o.team === MP.enemyTeam && o.connected)) {
    endGame(true, 'ALL OPPONENTS DISCONNECTED — DISTRICT SECURED');
  }
}

function onGameMsg(d, from) {
  const peer = peers.get(from);
  switch (d.t) {
    case 's': { // a player's state tick
      if (!peer) return;
      const st = peer.st;
      st.x = d.x; st.z = d.z; st.y = d.y; st.yaw = d.yaw;
      st.vx = d.vx; st.vz = d.vz; st.moving = !!d.m; st.age = 0;
      const e = peer.ent;
      if (e.alive && typeof d.hp === 'number') {
        e.hp = d.hp;
        e.bar.set(d.hp / e.maxHp);
      }
      for (const [id, yaw] of d.tu || []) {
        const t = netRegistry.get(id);
        if (t && t.alive && t.head && t.owner === from) { t.yaw = yaw; t.head.rotation.y = yaw; }
      }
      break;
    }
    case 'shot': // cosmetic: real damage arrives as 'hit'/'bhit' from the shooter
      spawnProjectile({
        pos: new THREE.Vector3(d.x, d.y, d.z),
        dir: new THREE.Vector3(d.dx, d.dy, d.dz),
        speed: d.s, damage: 0, team: d.tm, rocket: !!d.r, life: d.l, cosmetic: true,
      });
      break;
    case 'hit': { // a projectile hit an entity I own — I apply it
      const e = netRegistry.get(d.id);
      if (e && e.alive && e.owner === MP.playerId) damageEntity(e, d.d, peer ? peer.ent : null);
      break;
    }
    case 'bhit': { // base damage — shared entity, everyone mirrors the event
      const base = d.tm === 'blue' ? blueBase : redBase;
      if (base.alive) damageEntity(base, d.d, peer ? peer.ent : null);
      break;
    }
    case 'hp': { // authoritative hp of another player's entity after a hit
      const e = netRegistry.get(d.id);
      if (e && e.alive && e.owner === from) {
        e.hp = d.hp;
        if (e.bar) e.bar.set(e.hp / e.maxHp);
      }
      break;
    }
    case 'build': {
      if (!peer || netRegistry.has(d.id)) break;
      const t = makeTurretEntity(peer.team, d.x, d.z, d.id, from);
      t.remote = true;
      break;
    }
    case 'die': { // an entity died on its owner's client — mirror it
      const e = netRegistry.get(d.id);
      if (e && e.alive && e.owner === from) killEntity(e);
      break;
    }
    case 'fgrab': case 'fdrop': case 'fret': case 'fcap':
      onFlagMsg(d); // capture the flag — shared, unowned state
      break;
    case 'respawn': {
      if (!peer) return;
      const e = peer.ent;
      const sp = spawnPointFor(peer.team, peer.idx);
      e.alive = true;
      e.hp = e.maxHp;
      e.bar.set(1);
      e.y = groundHeightAt(sp.pos.x, sp.pos.z);
      e.vy = 0;
      Object.assign(peer.st, { x: sp.pos.x, z: sp.pos.z, y: e.y, vx: 0, vz: 0, moving: false, age: 0 });
      e.group.position.set(sp.pos.x, e.y, sp.pos.z);
      if (!entities.includes(e)) entities.push(e); // killEntity spliced it out
      scene.add(e.group);
      break;
    }
  }
}

/* ---------- per-frame: send my state, animate every replica ---------- */
let sendAcc = 0;
const SEND_DT = 1 / 15;

export function remoteUpdate(dt) {
  if (!MP.active) return;

  if (game.state !== 'menu') {
    sendAcc += dt;
    if (sendAcc >= SEND_DT) {
      sendAcc %= SEND_DT;
      sendState();
    }
  }

  const blink = Math.sin(game.elapsed * 10) > 0;
  for (const peer of peers.values()) {
    const e = peer.ent;
    if (!e.alive) continue;
    const st = peer.st;

    // ease toward the last packet, extrapolated briefly along its velocity
    st.age = Math.min(st.age + dt, 0.25);
    const tx = st.x + st.vx * st.age, tz = st.z + st.vz * st.age;
    const p = e.group.position;
    if (Math.hypot(tx - p.x, tz - p.z) > 14) { p.x = tx; p.z = tz; e.y = st.y; } // snap after teleports
    const k = 1 - Math.exp(-12 * dt);
    p.x += (tx - p.x) * k;
    p.z += (tz - p.z) * k;
    e.y += (st.y - e.y) * Math.min(1, 14 * dt);
    const dyaw = Math.atan2(Math.sin(st.yaw - e.yaw), Math.cos(st.yaw - e.yaw));
    e.yaw += dyaw * Math.min(1, 12 * dt);
    e.group.rotation.y = e.yaw;

    // a replica walks on the ground it actually covers, so legs stop the moment
    // the packets stop coming — a stale "moving" flag can't outlive it — and a
    // strafing peer shuffles sideways just like the pilot flying it sees
    const amp = animateWalk(e, dt, 9);
    // …but the owner's own velocity beats the one measured off the easing
    e.velX = st.vx;
    e.velZ = st.vz;
    p.y = e.y + Math.abs(Math.sin(e.walkPhase)) * 0.25 * amp;

    e.model.lampR.material.emissiveIntensity = blink ? 3 : 0.3;
    e.model.lampB.material.emissiveIntensity = blink ? 0.3 : 3;
  }
}

function sendState() {
  const p = player.group.position;
  const tu = [];
  for (const e of entities) {
    if (e.alive && e.kind === 'turret' && e.owner === MP.playerId && e.netId) {
      tu.push([e.netId, +e.yaw.toFixed(2)]);
    }
  }
  sendGame({
    t: 's',
    x: +p.x.toFixed(2), z: +p.z.toFixed(2), y: +player.y.toFixed(2),
    yaw: +player.yaw.toFixed(3),
    vx: +player.velX.toFixed(1), vz: +player.velZ.toFixed(1),
    hp: Math.round(player.hp),
    m: player.velX || player.velZ ? 1 : 0,
    tu,
  });
}

File: ios/MechVsMech/Engine/Remote.swift
import Foundation
import SceneKit
import UIKit
import simd

/* ============================================================
   Multiplayer match sync — ports systems/remote.js (up to 5 v 5)

   Ownership: each client simulates only what it owns — its player
   mech, the turrets it built, its projectiles. Everyone else's
   entities are replicas driven by network events (the server
   stamps each relayed event with the sender's playerId, so `from`
   is trustworthy):

     s        15 Hz state (position/yaw/velocity/hp + turret yaws)
     shot     a projectile was fired → spawn a cosmetic copy
     hit      my projectile hit an entity YOU own → you apply it
     hp       authoritative hp echo after a hit was applied
     bhit     damage to a base — shared, everyone mirrors it
     build    a turret was built
     die      an entity its owner simulates died → mirror it
     respawn  a player redeployed

   Damage to owned entities is shooter-reported but owner-applied,
   so hp has exactly one authority. Base hp converges because every
   client applies every bhit exactly once.
============================================================ */

struct PeerState {
    var x = 0.0, z = 0.0, y = 0.0, yaw = 0.0
    var vx = 0.0, vz = 0.0
    var moving = false
    var age = 0.0
}

final class Peer {
    let id: Int
    let name: String
    let team: Team
    let idx: Int
    let ent: Entity
    var connected = true
    var st = PeerState()

    init(id: Int, name: String, team: Team, idx: Int, ent: Entity) {
        self.id = id
        self.name = name
        self.team = team
        self.idx = idx
        self.ent = ent
    }
}

/* floating callsign so team fights stay readable */
private func makeNameTag(_ text: String, team: Team) -> SCNNode {
    let w = 256, h = 40
    UIGraphicsBeginImageContextWithOptions(CGSize(width: w, height: h), false, 1)
    let ctx = UIGraphicsGetCurrentContext()!
    let para = NSMutableParagraphStyle()
    para.alignment = .center
    let attrs: [NSAttributedString.Key: Any] = [
        .font: UIFont.systemFont(ofSize: 26, weight: .bold),
        .foregroundColor: UIColor(rgb: team == .red ? 0xffb3a6 : 0xa9c9ff),
        .paragraphStyle: para,
        .shadow: {
            let s = NSShadow()
            s.shadowColor = UIColor(white: 0, alpha: 0.9)
            s.shadowBlurRadius = 5
            return s
        }(),
    ]
    (text as NSString).draw(in: CGRect(x: 0, y: 4, width: w, height: h - 4), withAttributes: attrs)
    let img = UIGraphicsGetImageFromCurrentImageContext()!
    UIGraphicsEndImageContext()
    _ = ctx

    let plane = SCNPlane(width: 9, height: 9 * Double(h) / Double(w))
    let mat = SCNMaterial()
    mat.lightingModel = .constant
    mat.diffuse.contents = img
    mat.isDoubleSided = true
    mat.readsFromDepthBuffer = false
    mat.writesToDepthBuffer = false
    plane.firstMaterial = mat
    let node = SCNNode(geometry: plane)
    node.castsShadow = false
    node.renderingOrder = 10
    node.position.y = 9.6
    let bb = SCNBillboardConstraint()
    bb.freeAxes = .all
    node.constraints = [bb]
    return node
}

extension GameEngine {

    /* spawn a replica for every other player in the match */
    func initMatch() {
        guard let mp else { return }
        for p in mp.roster where p.id != mp.playerId {
            makePeer(p)
        }
    }

    private func makePeer(_ p: MPPlayer) {
        let idx = teamIndexOf(playerId: p.id, team: p.team, roster: mp!.roster)
        let sp = spawnPointFor(team: p.team, idx: idx)
        let y = level.groundHeightAt(sp.pos.x, sp.pos.z)
        let model = makeMechModel(p.team == .red ? RED_PAL : BLUE_PAL)
        let e = Entity(kind: .mech, team: p.team, node: model.group, hp: 300, hitRadius: 2.4, hitHeight: 7)
        e.netId = "player:\(p.id)"
        e.owner = p.id
        e.remote = true
        e.legL = model.legL
        e.legR = model.legR
        e.lampR = model.lampR
        e.lampB = model.lampB
        e.x = sp.pos.x
        e.z = sp.pos.z
        e.y = y
        e.yaw = atan2(sp.face.x - sp.pos.x, sp.face.z - sp.pos.z)
        e.bar = HealthBar(width: 5)
        e.syncNode()
        e.node.eulerAngles.y = Float(e.yaw)
        e.node.addChildNode(makeNameTag(p.name, team: p.team))
        registerEntity(e, barHeight: 8.2)

        let peer = Peer(id: p.id, name: p.name, team: p.team, idx: idx, ent: e)
        peer.st = PeerState(x: sp.pos.x, z: sp.pos.z, y: y, yaw: e.yaw,
                            vx: 0, vz: 0, moving: false, age: 0)
        peers[p.id] = peer
    }

    // MARK: - Inbound events (called on the render thread via enqueue)

    /* a relayed game event; `from` is the server-stamped sender playerId */
    func onGameMsg(_ d: [String: Any], from: Int) {
        enqueue { self.handleGameMsg(d, from: from) }
    }

    private func handleGameMsg(_ d: [String: Any], from: Int) {
        guard let t = jStr(d, "t") else { return }
        let peer = peers[from]
        switch t {
        case "s":   // a player's state tick
            guard let peer else { return }
            var st = peer.st
            st.x = jNum(d, "x") ?? st.x
            st.z = jNum(d, "z") ?? st.z
            st.y = jNum(d, "y") ?? st.y
            st.yaw = jNum(d, "yaw") ?? st.yaw
            st.vx = jNum(d, "vx") ?? 0
            st.vz = jNum(d, "vz") ?? 0
            st.moving = (jInt(d, "m") ?? 0) != 0
            st.age = 0
            peer.st = st
            let e = peer.ent
            if e.alive, let hp = jNum(d, "hp") {
                e.hp = hp
                e.bar?.set(hp / e.maxHp)
            }
            if let tu = d["tu"] as? [[Any]] {
                for pair in tu where pair.count == 2 {
                    guard let id = pair[0] as? String,
                          let yaw = (pair[1] as? NSNumber)?.doubleValue,
                          let tur = netRegistry[id], tur.alive, tur.head != nil, tur.owner == from
                    else { continue }
                    tur.yaw = yaw
                    tur.head?.eulerAngles.y = Float(yaw)
                }
            }

        case "shot":   // cosmetic: real damage arrives as hit/bhit from the shooter
            guard let team = Team(wire: jStr(d, "tm")),
                  let x = jNum(d, "x"), let y = jNum(d, "y"), let z = jNum(d, "z"),
                  let dx = jNum(d, "dx"), let dy = jNum(d, "dy"), let dz = jNum(d, "dz") else { return }
            spawnProjectile(
                pos: SIMD3(x, y, z), dir: SIMD3(dx, dy, dz),
                speed: jNum(d, "s") ?? 100, damage: 0, team: team,
                rocket: (jInt(d, "r") ?? 0) != 0, life: jNum(d, "l") ?? 3,
                src: nil, cosmetic: true)

        case "hit":   // a projectile hit an entity I own — I apply it
            guard let id = jStr(d, "id"), let e = netRegistry[id],
                  e.alive, e.owner == myPlayerId, let dmg = jNum(d, "d") else { return }
            damageEntity(e, dmg, src: peer?.ent)

        case "bhit":  // base damage — shared entity, everyone mirrors the event
            let base = jStr(d, "tm") == "blue" ? blueBase! : redBase!
            if base.alive, let dmg = jNum(d, "d") { damageEntity(base, dmg, src: peer?.ent) }

        case "hp":    // authoritative hp of another player's entity after a hit
            guard let id = jStr(d, "id"), let e = netRegistry[id],
                  e.alive, e.owner == from, let hp = jNum(d, "hp") else { return }
            e.hp = hp
            e.bar?.set(hp / e.maxHp)

        case "build":
            guard let peer, let id = jStr(d, "id"), netRegistry[id] == nil,
                  let x = jNum(d, "x"), let z = jNum(d, "z") else { return }
            let tur = makeTurretEntity(team: peer.team, x: x, z: z, netId: id, owner: from)
            tur.remote = true

        case "die":   // an entity died on its owner's client — mirror it
            guard let id = jStr(d, "id"), let e = netRegistry[id], e.alive, e.owner == from else { return }
            killEntity(e)

        case "respawn":
            guard let peer else { return }
            let e = peer.ent
            let sp = spawnPointFor(team: peer.team, idx: peer.idx)
            e.alive = true
            e.hp = e.maxHp
            e.bar?.set(1)
            e.y = level.groundHeightAt(sp.pos.x, sp.pos.z)
            e.vy = 0
            e.x = sp.pos.x
            e.z = sp.pos.z
            peer.st = PeerState(x: sp.pos.x, z: sp.pos.z, y: e.y, yaw: e.yaw,
                                vx: 0, vz: 0, moving: false, age: 0)
            e.syncNode()
            if !entities.contains(where: { $0 === e }) { entities.append(e) }  // killEntity spliced it
            if e.node.parent == nil { scene.rootNode.addChildNode(e.node) }

        default:
            break
        }
    }

    /* a player's socket dropped: despawn everything they owned — nobody is left
       to simulate it, and hits on it would never be applied */
    func handlePeerLeft(id: Int, name: String) {
        guard let peer = peers[id], peer.connected else { return }
        peer.connected = false
        for e in entities where e.owner == id {
            e.alive = false
            e.node.removeFromParentNode()
        }
        entities.removeAll { $0.owner == id && !$0.alive }
        delegate?.engineMessage("\(name.isEmpty ? peer.name : name) DISCONNECTED",
                                colorHex: peer.team == myTeam ? 0xff8a7a : 0x8ab4ff)
        if peer.team == enemyTeam
            && !peers.values.contains(where: { $0.team == enemyTeam && $0.connected }) {
            endGame(victory: true, reason: "ALL OPPONENTS DISCONNECTED — DISTRICT SECURED")
        }
    }

    /* a player reconnected mid-match (their client restarted from scratch) */
    func handlePeerJoined(id: Int, name: String) {
        guard let mp, let p = mp.roster.first(where: { $0.id == id }) else { return }
        if let existing = peers[id], existing.connected { return }
        // remove the stale replica, then build a fresh one
        if let old = peers[id] {
            old.ent.alive = false
            old.ent.node.removeFromParentNode()
            entities.removeAll { $0 === old.ent }
        }
        makePeer(p)
        delegate?.engineMessage("\(p.name) RECONNECTED", colorHex: 0x8ab4ff)
    }

    func handleConnectionLost() {
        endGame(victory: false, reason: "CONNECTION TO SERVER LOST")
    }

    // MARK: - Per-frame: send my state, ease every replica

    private static let sendDT = 1.0 / 15.0

    func remoteUpdate(dt: Double) {
        guard isMP else { return }

        if phase == .playing {
            sendAcc += dt
            if sendAcc >= Self.sendDT {
                sendAcc = sendAcc.truncatingRemainder(dividingBy: Self.sendDT)
                sendState()
            }
        }

        let blink = sin(elapsed * 10) > 0
        for peer in peers.values {
            let e = peer.ent
            if !e.alive { continue }
            var st = peer.st

            // ease toward the last packet, extrapolated briefly along its velocity
            st.age = min(st.age + dt, 0.25)
            let tx = st.x + st.vx * st.age, tz = st.z + st.vz * st.age
            if hypot(tx - e.x, tz - e.z) > 14 {   // snap after teleports
                e.x = tx; e.z = tz; e.y = st.y
            }
            let k = 1 - exp(-12 * dt)
            e.x += (tx - e.x) * k
            e.z += (tz - e.z) * k
            e.y += (st.y - e.y) * min(1, 14 * dt)
            let dyaw = angDiff(st.yaw, e.yaw)
            e.yaw += dyaw * min(1, 12 * dt)
            e.node.eulerAngles.y = Float(e.yaw)
            e.velX = st.vx
            e.velZ = st.vz
            peer.st = st

            if st.moving { e.walkPhase += dt * 9 }
            let sw = st.moving ? sin(e.walkPhase) * 0.55 : 0
            e.legL?.eulerAngles.x = Float(sw)
            e.legR?.eulerAngles.x = Float(-sw)
            e.syncNode(bob: st.moving ? abs(sin(e.walkPhase)) * 0.25 : 0)

            e.lampR?.emission.intensity = blink ? 3 : 0.3
            e.lampB?.emission.intensity = blink ? 0.3 : 3
        }
    }

    private func sendState() {
        var tu: [[Any]] = []
        for e in entities where e.alive && e.kind == .turret && e.owner == myPlayerId && e.netId != nil {
            tu.append([e.netId!, wire(e.yaw, 2)])
        }
        net?.sendGame([
            "t": "s",
            "x": wire(player.x, 2), "z": wire(player.z, 2), "y": wire(player.y, 2),
            "yaw": wire(player.yaw, 3),
            "vx": wire(player.velX, 1), "vz": wire(player.velZ, 1),
            "hp": Int(player.hp.rounded()),
            "m": (player.velX != 0 || player.velZ != 0) ? 1 : 0,
            "tu": tu,
        ])
    }
}
